### PR TITLE
[Bugfix] Fix Delete action for Exams/ExerciseGroups/Exercises

### LINKED ISSUE #1975: Text exercises with existing text clusters cannot be deleted
- [ ] `TextClusters` reference the exercise and need to be deleted first before the text exercise can be deleted.

Otherwise there will be a 

SQLIntegrityConstraintViolationException
Cannot delete or update a parent row: a foreign key constraint fails (`ArTEMiS`.`text_cluster`, CONSTRAINT `fk_text_block_exercise_id` FOREIGN KEY (`exercise_id`) REFERENCES `exercise` (`id`))

see https://sentry.io/organizations/ls1intum/issues/1807767491/?environment=prod&project=1440029&query=is%3Aunresolved&statsPeriod=1h

- [ ] Another issue happens when there are `result_ratings`:

Cannot delete or update a parent row: a foreign key constraint fails (`ArTEMiS`.`result_rating`, CONSTRAINT `fk_result_rating_result_id` FOREIGN KEY (`result_id`) REFERENCES `result` (`id`))

https://sentry.io/organizations/ls1intum/issues/1693664781/?end=2020-07-26T15%3A01%3A06&environment=prod&project=1440029&query=is%3Aunresolved&start=2020-07-26T14%3A55%3A06&utc=true

https://sentry.io/organizations/ls1intum/issues/1807805408/?end=2020-07-26T15%3A05%3A06&environment=prod&project=1440029&query=is%3Aunresolved&start=2020-07-26T15%3A03%3A06&utc=true

### LINKED ISSUE #1976: Quiz Exercises with short answer questions cannot be deleted
Some references are not cleared correctly, therefore the quiz cannot be deleted due to a

Cannot delete or update a parent row: a foreign key constraint fails (`ArTEMiS`.`short_answer_submitted_text`, CONSTRAINT `FKpkb6e1yjqhma5tgvabb9smyv4` FOREIGN KEY (`spot_id`) REFERENCES `short_answer_spot` (`id`) ON DELETE RESTRICT ON UPDATE RESTRICT)

see e.g. https://sentry.io/organizations/ls1intum/issues/1807782077/?end=2020-07-26T14%3A42%3A06&environment=prod&project=1440029&query=is%3Aunresolved&start=2020-07-26T14%3A20%3A06&utc=true

### PR DESCRIPTION
**Pull request by anditurdiu**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I documented the TypeScript code using JSDoc style.

### Motivation and Context
This PR is a follow-up to #1989 aimed at improving the delete functionality of exercise groups and exams. 
Deleting exams or exam elements does not work reliably. There have been several issues identified: 
Deleting exercises from exams is not possible, they are not removed. 
Ensure that all elements are deleted correctly (participations/build plans/repositories) etc. 
Fixes #1975
Fixes #1976

**Edit:** Additionally since it fits the functionality, I also refactored the delete function for courses in this PR. Firstly i moved the delete function to the service (similar to exercises and exams), and use the new `examService.deleteById()` to ensure that the exams are properly deleted. I also improved the documentation and enhanced the delete integration test.

### Description
- [x] Refactor delete exam & delete course, move to service
- [x] Fix #1975 by cascading the remove operation of text exercises. Extend integration test to ensure this.
- [x] #1976 could not be reproduced. We added an integration test to verify this and quiz exercises with short answer questions are deleted correctly. Therefore this issue can also be closed.
- [x] Improve warning message when the instructor is about to delete an exam or course. We now warn him that student exams, buildplans and repositories will also be deleted. 
- [x] If an exercise is now deleted which is part of existing student exams, we just remove the reference from the student exam. We do not delete the student exam. 
- [x] If you delete an exam, all buildplans & repositories are deleted.
- [x] If you want to delete an exercise group which contains a programming exercise, you are presented with the choice to also delete the buildplans & repositories. 
- [x] Create Integration test: `testDeleteExamWithStudentExamsAfterConductionAndEvaluation`. This exam contains all exercise types including all different quiz type questions (#1976).
- [x] Fix flaky StudentExam integration tests
- [x] Improve `testDeleteCourseWithPermission`
- [x] Improve Documentation

### Steps for Testing
_Delete Exam General_
1. Log in to Artemis
2. Navigate to Course Administration/Exams
3. Try deleting exercise groups with different exercises. 
4. Delete old exams and cleanup your Test Course :)
5. In case of programming exercises, double-check that buildplans & repositories are deleted as well
-----------------------------------------------------------------------------------------------
_Student Exams on delete exercise_
1. Create an exam with multiple exercises and generate the student exams
2. Delete an exercise or exercise group. 
3. Check that the student exams are not deleted and that they do not reference the deleted exercise anymore(It shouldn't show up).
------------------------------------------------------------------------------------------------
_Delete Course_
1. Delete one of the test courses, ideally, it should contain as many elements as possible. 
2. Everything should work as expected.

### CODE REVIEW

**Comment by anditurdiu** [comment]
@alexmardale @stefanwaldhauser thanks for the review. In my last commit I had introduced a bug, while trying to remove the expensive getter- This led to the component not detecting programming exercises in exercise groups correctly. This should be fixed now. As they are detected, the correct modal is shown and if you check the box the buildplans and repositories should be deleted.

File: src/test/java/de/tum/in/www1/artemis/QuizExerciseIntegrationTest.java [hunk 1]
```diff
@@ -31,14 +32,13 @@
 import de.tum.in.www1.artemis.domain.quiz.ShortAnswerQuestion;
 import de.tum.in.www1.artemis.domain.quiz.ShortAnswerSolution;
 import de.tum.in.www1.artemis.domain.quiz.ShortAnswerSpot;
-import de.tum.in.www1.artemis.repository.QuizExerciseRepository;
-import de.tum.in.www1.artemis.repository.ResultRepository;
-import de.tum.in.www1.artemis.repository.StudentParticipationRepository;
+import de.tum.in.www1.artemis.repository.*;
```
Comment (jpbernius) [inline src/test/java/de/tum/in/www1/artemis/QuizExerciseIntegrationTest.java:35]: Codacy found an issue: [Avoid unused imports such as 'de.tum.in.www1.artemis.repository'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968855)

File: src/main/java/de/tum/in/www1/artemis/service/ExamService.java [hunk 2]
```diff
@@ -19,10 +19,7 @@
 import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RequestBody;
 
-import de.tum.in.www1.artemis.domain.Exercise;
-import de.tum.in.www1.artemis.domain.ProgrammingExercise;
-import de.tum.in.www1.artemis.domain.Result;
-import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.domain.*;
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/ExamService.java:22]: Codacy found an issue: [Avoid unused imports such as 'de.tum.in.www1.artemis.domain'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968855)

File: src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java [hunk 3]
```diff
@@ -18,13 +19,12 @@
 import de.tum.in.www1.artemis.domain.enumeration.Language;
 import de.tum.in.www1.artemis.domain.exam.ExerciseGroup;
 import de.tum.in.www1.artemis.domain.participation.Participation;
-import de.tum.in.www1.artemis.repository.ExampleSubmissionRepository;
-import de.tum.in.www1.artemis.repository.ExerciseRepository;
-import de.tum.in.www1.artemis.repository.TextExerciseRepository;
-import de.tum.in.www1.artemis.repository.TextSubmissionRepository;
+import de.tum.in.www1.artemis.repository.*;
```
Comment (jpbernius) [inline src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java:22]: Codacy found an issue: [Avoid unused imports such as 'de.tum.in.www1.artemis.repository'](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968855)

File: src/test/java/de/tum/in/www1/artemis/StudentExamIntegrationTest.java [hunk 4]
```diff
@@ -1003,5 +1003,81 @@ else if (quizQuestion instanceof MultipleChoiceQuestion) {
                 request.putWithResponseBody("/api/exercises/" + exercise.getId() + "/submissions/exam", quizSubmission, QuizSubmission.class, HttpStatus.OK);
             }
         }
+        return studentExam;
+    }
+
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    public void testDeleteExamWithStudentExamsAfterConductionAndEvaluation() throws Exception {
```
Comment (jpbernius) [inline src/test/java/de/tum/in/www1/artemis/StudentExamIntegrationTest.java:1011]: Codacy found an issue: [JUnit tests should include assert() or fail()](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968855)

File: src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java [hunk 5]
```diff
@@ -73,13 +79,17 @@ public void submitEnglishTextExercise() throws Exception {
 
     @Test
     @WithMockUser(value = "instructor1", roles = "INSTRUCTOR")
-    public void deleteTextExerciseWithSubmissionWithTextBlocks() throws Exception {
+    public void deleteTextExerciseWithSubmissionWithTextBlocksAndClusters() throws Exception {
```
Comment (jpbernius) [inline src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java:82]: Codacy found an issue: [JUnit tests should include assert() or fail()](https://app.codacy.com/gh/ls1intum/Artemis/pullRequest?prid=5968855)

File: src/main/webapp/i18n/en/course.json [hunk 6]
```diff
@@ -11,7 +11,7 @@
             "updated": "A Course was updated with title {{ param }}",
             "deleted": "A Course was deleted with title {{ param }}",
             "delete": {
-                "question": "Are you sure you want to permanently delete the course <strong>{{ title }}</strong>? All associated elements will be deleted, including the exams. This action can NOT be undone!",
+                "question": "Are you sure you want to permanently delete the course <strong>{{ title }}</strong>? All associated elements will be deleted, including the exams, exercises, repositories and buildplans. This action can NOT be undone!",
```
Comment (krusche) [inline src/main/webapp/i18n/en/course.json:14]:
```suggestion
                "question": "Are you sure you want to permanently delete the course <strong>{{ title }}</strong>? All associated elements will be deleted, including the exams, exercises, repositories and build plans. This action can NOT be undone!",
```

Comment (anditurdiu) [reply src/main/webapp/i18n/en/course.json]: See f9e39e2dbe10ffa13483affe2de6359ac2244213

File: src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java (hunk 5 above)
Comment (anditurdiu) [reply src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java:85]: See f9e39e2dbe10ffa13483affe2de6359ac2244213

File: src/test/java/de/tum/in/www1/artemis/StudentExamIntegrationTest.java (hunk 4 above)
Comment (anditurdiu) [reply src/test/java/de/tum/in/www1/artemis/StudentExamIntegrationTest.java:1011]: See f9e39e2dbe10ffa13483affe2de6359ac2244213

File: src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java (hunk 3 above)
Comment (anditurdiu) [reply src/test/java/de/tum/in/www1/artemis/TextExerciseIntegrationTest.java]: See See f9e39e2dbe10ffa13483affe2de6359ac2244213

File: src/main/java/de/tum/in/www1/artemis/service/ExamService.java (hunk 2 above)
Comment (anditurdiu) [reply src/main/java/de/tum/in/www1/artemis/service/ExamService.java]: See f9e39e2dbe10ffa13483affe2de6359ac2244213

File: src/test/java/de/tum/in/www1/artemis/QuizExerciseIntegrationTest.java (hunk 1 above)
Comment (anditurdiu) [reply src/test/java/de/tum/in/www1/artemis/QuizExerciseIntegrationTest.java]: See f9e39e2dbe10ffa13483affe2de6359ac2244213

**Pull request merged**